### ISSUE TITLE
Unreliable free memory resulting in models not running

### ISSUE THREAD

**Issue by ddpasa** [ollama/ollama#6918]
### What is the issue?

From what I understand, new versions of ollama compare the expected memory requirements of a model with the amount of free memory seen by ollama, and prints an error message if the model memory requirements are larger.  This make a lof of sense.

However, the free memory on Linux is (from what I understand) is not a very reliable estimate.  For the same model on the same machine, I have had cases where ollama ran successfully, or reported insufficient memory.

Is it possible to disable this feature entirely?

### OS

Linux

### Ollama version

latest mainline

**Comment by rick-github** [comment]
Broadly speaking, ollama wants the sum of unallocated RAM and unallocated swap to be more than the required memory for loading the model + context space that don't fit on the GPU.  The [server logs](https://github.com/ollama/ollama/blob/main/docs/troubleshooting.md#how-to-troubleshoot-issues) will show the relevant values.  If you are finding that a model loads sometimes and not others, then ollama thinks that your system is close to over-committing RAM and doesn't want to get in to a situation where the OOM-killer starts sniping processes.  You can check the figures in the logs and if you find that the data is inconsistent then that should be followed up.  You can mitigate the problems with model loading by using a smaller model, setting a smaller context size, or adding swap.

**Comment by ddpasa** [comment]
> If you are finding that a model loads sometimes and not others, then ollama thinks that your system is close to over-committing RAM and doesn't want to get in to a situation where the OOM-killer starts sniping processes.

I think this is exactly what is happening.

I have 16GB of ram on my laptop, and the free memory ollama sees fluctuated between 9.5GB and all the way up to 15GB.  This is a huge range of memory fluctuations.

**Comment by ddpasa** [comment]
I think the current logic is a safe convervative choice that works most of the time.  However, I know my own system very well, and would like to override the available memory with a larger value for making sure I avoid changing behaviour.

**Comment by dhiltgen** [comment]
We look at available memory which should be buffer cache aware, along with swap free space to establish a threshold so we can block model loads that exceed that.  Can you describe your scenario a bit more?  Are you loading different models in rapid fire where we unload one to make room for the next, but still think there isn't room due to stale memory information?  For GPUs we wait up to 5s for the VRAM reporting to converge, but we don't currently have code in place to do that for system memory.  If that's the scenario you're running into, maybe that enhancement would help address the problem.

**Comment by ddpasa** [comment]
> We look at available memory which should be buffer cache aware, along with swap free space to establish a threshold so we can block model loads that exceed that. Can you describe your scenario a bit more? Are you loading different models in rapid fire where we unload one to make room for the next, but still think there isn't room due to stale memory information? For GPUs we wait up to 5s for the VRAM reporting to converge, but we don't currently have code in place to do that for system memory. If that's the scenario you're running into, maybe that enhancement would help address the problem.

I don't think there is a major issue with the current logic, but it does not work well in case where I really want to push my system and run the largest model I can get away with.  It seems a little too conservative.

I have a suspicion that it's due to other programs in my laptop using memory that confuses ollama.

A very simple solution is to allow users to override this value with an environment variable.  It keeps the default safe behaviour, while allowing us to run with large models right around the memory threshold.

**Comment by nielsproest** [comment]
I also experience this if i disable my swap.
Im running ZFS though, so it might not calculate from memory used as cache.

**Comment by ddpasa** [comment]
I think overriding with a environment variable ist he cleanest solution.

**Comment by rick-github** [comment]
Does adding swap not solve the problem?

**Comment by xgdgsc** [comment]
https://discussions.apple.com/thread/7417584?answerId=7417584021&sortBy=rank#7417584021 seems macos has no option of manually add. So considering both macos and Windows manages swap dynamically by default, adding an env variable should work.

**Comment by rick-github** [comment]
```
OLLAMA_MEM=30000000000 perl -e '$x="a" x $ENV{OLLAMA_MEM}; exec ("ollama","serve);'
```

**Comment by rick-github** [comment]
Python is more likely to be installed on a windows system than perl, so for better cross platform support:
```python
#!/usr/bin/env python3

# For systems with dynamic swap, alloc a large buffer to force the
# system to add swap, then exec ollama into that free space.
# Install this in a path before the actual ollama binary, and
# adjust `ollama_binary` below to point to the real ollama.
#
# use:  OLLAMA_MEMORY=10000000000 ollama run some-large-model

import os
import platform
import sys

ollama_binary = "ollama"
if platform.system() == "Windows":
  ollama_binary = "ollama.exe"

_ = "a"*int(os.environ.get("OLLAMA_MEMORY",0))
os.execvp(ollama_binary, [ollama_binary]+sys.argv[1:])
```

**Comment by xgdgsc** [comment]
Thanks. Works for me.

**Comment by unicorn667** [comment]
worked not for me

**Comment by thojo0** [comment]
For me I also can't load models even if enough memory is available.

> We look at available memory which should be buffer cache aware, along with swap free space to establish a threshold so we can block model loads that exceed that.

I don't think, this is working correctly under linux (ollama v0.18.0): `free -h`
```
               total        used        free      shared  buff/cache   available
Mem:            32Gi        37Mi        10Gi        64Ki        21Gi        31Gi
Swap:          512Mi       4.4Mi       507Mi
```
In this case ollama refuses to load models bigger than 10G, but should be able to load models up to 31G.

**Comment by aldem** [comment]
Just hit similar issue - it reports that I have "not enough" RAM.

Checking `MemFree` is *not reliable* (even without containers), because `MemFree` *does not* reflect *available* memory, which is reported as `MemAvailable`:

```
MemAvailable %lu (since Linux 3.14)
  An estimate of how much memory is available for starting new applications, without swapping.
```

`MemFree` only shows *not used* memory, this means in particular that if most of the RAM is used for caches then it will be low, like in my case:

```
$ free -h
               total        used        free      shared  buff/cache   available
Mem:            64Gi        30Gi       557Mi       4.7Gi        37Gi        33Gi
Swap:             0B          0B          0B
```

I have 33G available but `ollama` refuses to run with message `Error: model requires more system memory (539.6 MiB) than is available (486.9 MiB)`.

**Comment by rick-github** [comment]
> Checking MemFree is not reliable (even without containers), because MemFree does not reflect available memory, which is reported as MemAvailable:

ollama uses MemAvailable.  [Server logs](https://docs.ollama.com/troubleshooting) will aid in debugging.

**Comment by aldem** [comment]
I am not sure it does, because (v0.19.0):

```
time=2026-03-31T01:37:08.450+02:00 level=INFO source=sched.go:484 msg="system memory" total="64.0 GiB" free="2.4 GiB" free_swap="0 B"
``` 
Shows free 2.4 GiB, which matches meminfo:

```
MemTotal:       67108864 kB
MemFree:         2283532 kB
MemAvailable:   34857276 kB
```

Besides, as I mentioned above, it refused to run with 33 GiB available.

**Comment by rick-github** [comment]
https://github.com/ollama/ollama/blob/31f968fe1f0f774fe20ee0c64f749e90d54147fd/discover/cpu_linux.go#L43

[Server logs](https://docs.ollama.com/troubleshooting) will aid in debugging.

**Comment by aldem** [comment]
I saw the code, but... Debugging. Ok. Lets see:

```
time=2026-03-31T02:04:14.212+02:00 level=INFO source=sched.go:484 msg="system memory" total="64.0 GiB" free="131.0 MiB" free_swap="0 B"
...
time=2026-03-31T02:04:14.212+02:00 level=WARN source=server.go:1046 msg="model request too large for system" requested="330.0 MiB" available="131.0 MiB" total="64.0 GiB" free="131.0 MiB" swap="0 B"
time=2026-03-31T02:04:14.212+02:00 level=INFO source=sched.go:511 msg="Load failed" model=/home/ollama/.ollama/models/blobs/sha256-33a8a1b6a1cbba662f292d32bb55f8d109c0e6cb02de2d243a1b70705ea20986 error="model requires more system memory (330.0 MiB) than is available (131.0 MiB)"
```

My cache is full, and `meminfo` is:

```
$ egrep Mem /proc/meminfo
MemTotal:       67108864 kB
MemFree:          131010 kB
MemAvailable:   35081704 kB
```

Does it look that it actually uses `MemAvailable`?

**Comment by rick-github** [comment]
> Does it look that it actually uses MemAvailable?

Hard to say.  If only there were [server logs](https://docs.ollama.com/troubleshooting) to look at.

**Comment by aldem** [comment]
Sorry, which *else* server logs should I post? Or you mean that I have to post *everything* from the log, even unrelated to memory?

**Comment by rick-github** [comment]
Server logs contains information about environment, device selection, model parameters, layer allocation etc which may or not be useful in debugging.  Better to provide a full log that may have too much detail than 3 lines of log which may exclude relevant information.

**Comment by aldem** [comment]
OK, the full log: https://gist.github.com/aldem/f3a3313d89fe83f8dbeb05564bcfec87

And failed attempt:
```
$ ./ollama run ordis/jina-embeddings-v2-base-code "some text here"; egrep Mem /proc/meminfo
Error: model requires more system memory (330.0 MiB) than is available (170.3 MiB)
MemTotal:       67108864 kB
MemFree:          154768 kB
MemAvailable:   35162983 kB
```

While the code should work, it doesn't... free memory in log is matching `MemFree` (with slight difference).

I am running it in LXC container (Proxmox 9, Debian 13) - not sure if this matters.

**Comment by rick-github** [comment]
What does the `egrep` return when it's run inside the LXC container?

**Comment by aldem** [comment]
`ollama serve` and `ollama run ...` with `egrep` are running in the same container, so it returns actual container data.

**Comment by JordanLoehr** [comment]
I noticed this too running ollama on k3s on a raspberry pi 5 (8gb) for testing. 

Would get
`model requires more system memory (7.3 GiB) than is available (2.4 GiB)`

Despite /proc/meminfo showing:
`MemAvailable:    7434880 kB`

Looking at https://github.com/ollama/ollama/blob/main/discover/cpu_linux.go though, its not just always looking at MemAvailable, but it also does a pass that checks if it is in a cgroup (`getCPUMemByCgroups(mem)`) and overrides the MemAvailable with the cgroup values if present. 

https://github.com/ollama/ollama/blob/8c8f8f3450d39735355fc6cd7f2e436c8aa42ab1/discover/cpu_linux.go#L69-L79

The problem with this is `"/sys/fs/cgroup/memory.current"` includes things such as the page cache, which MemAvailable doesn't.

On my system running in k3s this explains the discrepancy:

`/sys/fs/cgroup/memory.max` returns `max`, which causes getUint64ValueFromFile to error and keep the total from /proc/meminfo instead (8256640 kB), but `/sys/fs/cgroup/memory.current` returns `5825003520` (bytes, or 5825004 kB),

So  8256640-5825004 = 2431636 or 2.4GiB, which is what the original error is showing as free, even though MemAvailable is showing 7.4GiB free.

To get the cgroup value closer to what MemAvailable is, you have to subtract all the reclaimable memory from memory.current, most of which you can get from memory.stat.

eg:

available = total - (memory.current - (memory.stat.anon + memory.stat.kernel + memory.stat.slab_unreclaimable + memory.stat.kernel_stack + memory.stat.pagetables + memory.stat.sec_pagetables + memory.stat.sock + memory.stat.vmalloc) 

However this isn't exactly the same as MemAvailable because that factors in some of the per zone low watermark values from /proc/zoneinfo. Or just copy what Kubernetes does https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#memory-signals and use `total - (memory.current - memory.stat.inactivefile)` to get close enough.

tl;dr: if you are running in a cgroupv2 it's using the value from memory.current, not /proc/meminfo MemAvailable, to calculate the free memory, with the former including the page cache and other values.

**Comment by aldem** [comment]
@JordanLoehr Exactly, you have nailed it! 👍

**Comment by mirceanis** [comment]
I'm facing a similar issue.
Running ollama in a container.
On 32GB RAM + 8GB VRAM I can run gemma4:26b-a4b @ 4 bit quantization with a 192k context window.
BUT, as soon as the model is shut down it won't restart

**Comment by markasoftware-tc** [comment]
Yep I believe my pr #13782 fixes this exact issue

**Comment by dhiltgen** [comment]
This should be fixed now in the latest versions, or at least materially better.  If you still have problems with the latest version, please share an updated server log with OLLAMA_DEBUG=1 set and details about the repro scenario, OS, GPU/driver, etc. and I can reopen the issue.

**Comment by markasoftware-tc** [comment]
@dhiltgen this issue is not resolved. The memory accounting still does not subtract page cache from used memory, as #13782 and #15701 do. Please re-open the issue.

This is not the sort of issue where the fix should make things "at least materially better". It's a true bug in memory accounting (ie, not due to actual high memory usage) with a simple black-and-white fix that fully resolves it.

**Comment by markasoftware-tc** [comment]
ah, I see that the exact error message complained about here can't trigger any more because things have been reorganized a bit.

But the free memory calculation is still wrong, and is used in a few other places, like checking whether it needs to evict existing models before loading a new model, and whether `disableMmapForHostPressure` is activated. Whatever rearchitecting has happened simply changes this from being a highly visible bug that prevents model loading to being an invisible bug that causes non-optimal behavior in containers in certain cases
